javascript/ql/lib/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -1992,20 +1992,26 @@ module PathGraph {
 }
 
 /**
- * Gets an operand of the given `&&` operator.
- *
- * We use this to construct the transitive closure over a relation
- * that does not include all of `BinaryExpr.getAnOperand`.
+ * Gets a logical `and` expression, or parenthesized expression, that contains `guard`.
  */
-private Expr getALogicalAndOperand(LogAndExpr e) { result = e.getAnOperand() }
+private Expr getALogicalAndParent(BarrierGuardNode guard) {
+  barrierGuardIsRelevant(guard) and result = guard.asExpr()
+  or
+  result.(LogAndExpr).getAnOperand() = getALogicalAndParent(guard)
+  or
+  result.getUnderlyingValue() = getALogicalAndParent(guard)
+}
 
 /**
- * Gets an operand of the given `||` operator.
- *
- * We use this to construct the transitive closure over a relation
- * that does not include all of `BinaryExpr.getAnOperand`.
+ * Gets a logical `or` expression, or parenthesized expression, that contains `guard`.
  */
-private Expr getALogicalOrOperand(LogOrExpr e) { result = e.getAnOperand() }
+private Expr getALogicalOrParent(BarrierGuardNode guard) {
+  barrierGuardIsRelevant(guard) and result = guard.asExpr()
+  or
+  result.(LogOrExpr).getAnOperand() = getALogicalOrParent(guard)
+  or
+  result.getUnderlyingValue() = getALogicalOrParent(guard)
```

Comment:
can you confirm: this line is the reason you made this new predicate, rather than continuing to use `getALogicalOrOperand+`? It makes it so that even barrier guard nodes that are wrapped in (parens) are handled properly?

Reply:
Yes.   

I could also have written a simpler step predicate like: 
```CodeQL
Expr getLogicalAndThing(Expr prev) {
  result = prev.(LogAndExpr).getAnOperand()
  or
  result = prev.getUnderlyingValue()
}
```

And then used that like `getLogicalAndThing+(returnExpr) = guard`.   
Which would be more in line with the previous implementation.  

--- 

However, the approach I have chosen gives better performance, because I'm limiting the base-case, which means that the transitive closure gets way smaller.  
(Using the `+`/`'*` operators is generally faster, but it sometimes computes some very large intermediate results if it doesn't limit the base case).

Reply:
Thanks for the thorough explanation <3